google-cloud-spanner/lib/google/cloud/spanner/client.rb

```diff
@@ -728,7 +731,11 @@ def transaction deadline: 120, &block
           start_time = current_time
 
           @pool.with_transaction do |tx|
+            unless Thread.current[:transaction_id].nil?
+              fail "Nested transactions are not allowed"
+            end
             begin
+              Thread.current[:transaction_id] = tx.transaction_id
```

Comment:
It currently will allow nested snapshots. I can add the same logic there as well.